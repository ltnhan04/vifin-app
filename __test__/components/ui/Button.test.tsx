jest.useFakeTimers();

import React from "react";
import { render, fireEvent, cleanup } from "@testing-library/react-native";
import ButtonSubmit from "@/components/ui/Button";

describe("ButtonSubmit Component", () => {
  const title = "Submit";
  const background = "#123456";
  const textColor = "#FFFFFF";
  const handleOnPress = jest.fn();
  const iconMock = { uri: "dummy-icon.png" };

  afterEach(cleanup);

  it("renders correctly when not loading and not disabled", () => {
    const rendered = render(
      <ButtonSubmit
        title={title}
        isLoading={false}
        isDisabled={false}
        background={background}
        textColor={textColor}
        handleOnPress={handleOnPress}
      />
    );

    // Kiểm tra hiển thị title
    expect(rendered.getByText(title)).toBeTruthy();
    // ActivityIndicator không xuất hiện
    expect(rendered.queryByTestId("activity-indicator")).toBeNull();

    // Kiểm tra rằng button không bị disabled
    const button = rendered.getByTestId("button-submit");
    // Kiểm tra thông qua accessibilityState.disabled
    expect(button.props.accessibilityState?.disabled).toBeFalsy();

    expect(rendered.toJSON()).toMatchSnapshot();
  });

  it("renders correctly when loading: shows ActivityIndicator and disables button", () => {
    const rendered = render(
      <ButtonSubmit
        title={title}
        isLoading={true}
        isDisabled={false}
        background={background}
        textColor={textColor}
        handleOnPress={handleOnPress}
      />
    );

    // Kiểm tra hiển thị title
    expect(rendered.getByText(title)).toBeTruthy();
    // ActivityIndicator phải xuất hiện
    const activityIndicator = rendered.getByTestId("activity-indicator");
    expect(activityIndicator).toBeTruthy();

    // Button phải bị disable
    const button = rendered.getByTestId("button-submit");
    expect(button.props.accessibilityState?.disabled).toBeTruthy();
  });

  it("renders correctly when disabled: button disabled and no ActivityIndicator", () => {
    const rendered = render(
      <ButtonSubmit
        title={title}
        isLoading={false}
        isDisabled={true}
        background={background}
        textColor={textColor}
        handleOnPress={handleOnPress}
      />
    );

    // Kiểm tra hiển thị title
    expect(rendered.getByText(title)).toBeTruthy();
    // ActivityIndicator không xuất hiện
    expect(rendered.queryByTestId("activity-indicator")).toBeNull();

    // Button bị disabled
    const button = rendered.getByTestId("button-submit");
    expect(button.props.accessibilityState?.disabled).toBeTruthy();
  });

  it("renders icon if provided", () => {
    const rendered = render(
      <ButtonSubmit
        title={title}
        isLoading={false}
        isDisabled={false}
        background={background}
        textColor={textColor}
        icon={iconMock}
        handleOnPress={handleOnPress}
      />
    );

    const image = rendered.getByTestId("button-icon");
    expect(image.props.source).toEqual(iconMock);
  });

  it("calls handleOnPress when button is pressed", () => {
    const rendered = render(
      <ButtonSubmit
        title={title}
        isLoading={false}
        isDisabled={false}
        background={background}
        textColor={textColor}
        handleOnPress={handleOnPress}
      />
    );

    const button = rendered.getByTestId("button-submit");
    fireEvent.press(button);
    expect(handleOnPress).toHaveBeenCalledTimes(1);
  });
});
