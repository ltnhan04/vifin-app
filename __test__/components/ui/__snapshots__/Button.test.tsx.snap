// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonSubmit Component renders correctly when not loading and not disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#123456",
      "opacity": 1,
    }
  }
  testID="button-submit"
>
  <View
    className="flex flex-row items-center justify-center gap-x-4"
  >
    <Text
      className="text-center font-rubik-medium text-lg"
      style={
        {
          "color": "#FFFFFF",
        }
      }
    >
      Submit
    </Text>
  </View>
</View>
`;
